app_llm.py

```diff
@@ -380,25 +385,24 @@ def set_first_row_as_header(df):
                     # Add column functionality
                     # new_col_name = st.text_input(f"New column name", key=f"new_col_{i}")
                     if st.button(f"Add Column", key=f"add_col_{i}"):
-                        table_dfs[i][int(table_dfs[i].columns[-1]) + 1] = None
-
+                        table_dfs[i][str(int(table_dfs[i].columns[-1]) + 1)] = None
+                        save_st_table(table_dfs)
+        
                     # Delete column functionality
                     if not st.session_state.table_dfs[i].empty:
                         col_to_delete = st.selectbox(f"Column to delete", st.session_state.table_dfs[i].columns,
                                                     key=f"del_col_{i}")
                         if st.button(f"Delete Column", key=f"delete_col_{i}"):
                             table_dfs[i] = table_dfs[i].drop(columns=[col_to_delete])
-            
+                            save_st_table(table_dfs)
+
             # This can normalize table headers to match DHIS2 using Levenstein distance or semantic search
             # TODO: Currently there's only a small set of hard coded fields, which might look weird to the user, so it's left of for the demo
             #if st.button(f"Correct field names", key=f"correct_names"):
             #     table_dfs = correct_field_names(table_dfs)
-                
-            # Rerun the code to display any edits made by user
-            for idx, table in enumerate(table_dfs):
-                if not table_dfs[idx].equals(st.session_state.table_dfs[idx]):
-                    st.session_state.table_dfs = table_dfs
-                    st.rerun()
+            if st.button("Save changes", type="primary"):    
```

Comment:
I think the text on this button might be a little confusing. I know you're doing it to save the state, but from the user's perspective, they can already see the changes to the table. Does saving it mean pushing to DHIS2 or something else? 

When is it important to the functionality for DHIS2 upload to save the state? Maybe we could just have the state be changed when the user is ready to display the key/values for upload?

Reply:
The changes made by the user in the streamlit editor needs to be saved before calling the correct_field_names function so that any text editing the user has done will be user for similarity check.

So the ideal order is, 

- User makes changes in tables
- **Save changes**
- Correct field names 
   - Make any more changes if required
   - **Save changes**
- Generate key value pairs
- Upload ot DHIS2

Reply:
So any time you edit the table manually, save your changes